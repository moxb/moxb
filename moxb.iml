<?xml version="1.0" encoding="UTF-8"?>
<module type="WEB_MODULE" version="4">
  <component name="NewModuleRootManager" inherit-compiler-output="true">
    <exclude-output />
    <content url="file://$MODULE_DIR$">
      <sourceFolder url="file://$MODULE_DIR$/packages/antd/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/packages/meteor/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/packages/moxb/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/packages/react-html/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/examples/src" isTestSource="false" />
      <excludeFolder url="file://$MODULE_DIR$/packages/moxb/dist" />
      <excludeFolder url="file://$MODULE_DIR$/packages/meteor/dist" />
      <excludeFolder url="file://$MODULE_DIR$/packages/react-html/dist" />
      <excludeFolder url="file://$MODULE_DIR$/packages/antd/dist" />
      <excludeFolder url="file://$MODULE_DIR$/admin/node-installation" />
      <excludeFolder url="file://$MODULE_DIR$/admin/bin-tools" />
      <excludeFolder url="file://$MODULE_DIR$/examples/dist" />
      <excludeFolder url="file://$MODULE_DIR$/examples/.cache" />
      <excludeFolder url="file://$MODULE_DIR$/coverage" />
    </content>
    <orderEntry type="inheritedJdk" />
    <orderEntry type="sourceFolder" forTests="false" />
  </component>
</module>