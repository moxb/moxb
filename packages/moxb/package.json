{
  "name": "@moxb/moxb",
  "version": "0.2.0-beta.59",
  "description": "MobX data binding",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "main": "dist",
  "typings": "dist/",
  "files": [
    "dist/**/*"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/moxb/moxb.git"
  },
  "author": "moxb (moxb.org)",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/moxb/moxb/issues"
  },
  "homepage": "https://github.com/moxb/moxb#readme",
  "keywords": [
    "mobx",
    "UI"
  ],
  "jest": {
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "jsx",
      "json",
      "node"
    ],
    "transform": {
      "^.+\\.(ts|tsx)$": "ts-jest"
    },
    "testEnvironment": "node",
    "testMatch": [
      "**/src/**/__tests__/*.(ts|tsx)"
    ],
    "modulePaths": [
      "<rootDir>/peer-dependencies/node_modules"
    ]
  },
  "peerDependencies": {
    "mobx": "^5.15.6",
    "tslib": "^2.0.1"
  },
  "devDependencies": {
    "@types/atob": "2.1.2",
    "@types/jest": "26.0.13",
    "@types/lodash.flatten": "4.4.6",
    "@types/urijs": "1.19.11",
    "mobx": "5.13.1",
    "tslib": "1.10.0"
  },
  "dependencies": {
    "btoa-lite": "1.0.0",
    "history": "5.0.0",
    "lodash.flatten": "4.4.0",
    "moment": "2.29.1",
    "urijs": "1.19.2"
  }
}
