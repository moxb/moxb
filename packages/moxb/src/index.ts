export * from './impl';
export * from './types';
export * from './routing';
