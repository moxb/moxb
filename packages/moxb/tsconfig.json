{
    "compilerOptions": {
        "module": "commonjs",
        "target": "ES5",
        "jsx": "react",
        "lib": ["dom","es6","es2015","es2017", "dom.iterable","scripthost"],
        "inlineSourceMap": true,
        "importHelpers":true,
        "experimentalDecorators": true,
        "noImplicitAny": true,
        "noImplicitThis": true,
        "strict": true,
        "outDir": "./dist",
        "declaration": true,
        "skipLibCheck": true,
        "strictNullChecks": true,
        "removeComments": false,
        "noUnusedLocals": true,
        "noUnusedParameters": true,
        "allowUnreachableCode": false,
        "baseUrl": "./src",
        "paths": {
            "*": [
                "*",
                "../../../node_modules/*"
            ]
        }
    },
    "exclude": [
        "./node_modules",
        "./dist"
    ]
}
