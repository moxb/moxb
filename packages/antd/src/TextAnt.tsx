import { Action, t, Text } from '@moxb/moxb';
import { Button, Icon, Input } from 'antd';
import { InputProps, SearchProps } from 'antd/lib/input';
import { observer } from 'mobx-react';
import * as React from 'react';
import { CSSProperties } from 'react';
import { BindAntProps, parseProps } from './BindAnt';
import { BindFormItemAntProps, FormItemAnt, parsePropsForChild } from './FormItemAnt';

export interface BindStringAntProps extends BindAntProps<Text>, InputProps {
    operation: Text;
    useDoubleClickToEdit?: boolean;
    help?: string;
    onPressEnter?(): void;
    rows?: number;
}

export interface BindSearchStringAntProps extends SearchProps {
    operation: Text;
    searchAction: Action;
    enterButton?: string;
    clearbuttonstyle?: {};
    style?: {};
    btnText?: string;
}

@observer
export class TextAnt extends React.Component<BindStringAntProps> {
    // tslint:disable-next-line:cyclomatic-complexity
    render() {
        const { operation, id, inputType, value, invisible, ...props } = parseProps(this.props, this.props.operation);
        if (invisible) {
            return null;
        }
        if ((inputType || operation.control) === 'textarea') {
            return (
                <Input.TextArea
                    id={id}
                    placeholder={operation.placeholder}
                    onFocus={operation.onEnterField}
                    onBlur={operation.onExitField}
                    value={operation.value || value || ''}
                    onChange={(e: any) => operation.setValue(e.target.value)}
                    {...props as any}
                />
            );
        } else {
            return (
                <Input
                    id={id}
                    placeholder={operation.placeholder}
                    onFocus={operation.onEnterField}
                    onBlur={operation.onExitField}
                    value={operation.value || value || ''}
                    onChange={(e: any) => operation.setValue(e.target.value)}
                    {...props as any}
                />
            );
        }
    }
}

/**
 * Use `onSearch` for the search function! Then it will work with the Enter as well as with the search button!
 */
@observer
export class TextSearchAnt extends React.Component<BindSearchStringAntProps> {
    private clearBtnOffset: string;

    constructor(props: BindSearchStringAntProps) {
        super(props);
        this.clearBtnOffset = '85px';
    }

    componentDidMount(): void {
        this.getSearchBtnWidth();
    }

    getSearchBtnWidth() {
        // The search button can change in size, so we adjust the clear btn position
        const elements = document.getElementsByClassName('ant-input-search-button');
        const requiredElement = elements[0] as HTMLElement;
        this.clearBtnOffset = requiredElement ? requiredElement.offsetWidth + 10 + 'px' : '85px';
    }

    handleKeyDown(e: React.KeyboardEvent<HTMLElement>, inputValue?: Text) {
        if (e.keyCode === 27 && inputValue) {
            inputValue.setValue('');
        }
    }

    render() {
        const Search = Input.Search;
        const { operation, id, value, invisible, style, enterButton, searchAction, ...props } = parseProps(
            this.props,
            this.props.operation
        );
        if (invisible) {
            return null;
        }
        const clearButtonStyle: CSSProperties = this.props.clearbuttonstyle || {
            position: 'absolute',
            display: 'block',
            right: this.clearBtnOffset,
            height: '24px',
            width: '24px',
            borderRadius: '20px',
            padding: '0px',
            top: '4px',
            backgroundColor: 'lightgrey',
            zIndex: 1,
        };

        return (
            <div style={{ ...style, ...{ position: 'relative', marginBottom: '1.5em' } }}>
                <Search
                    id={id}
                    placeholder={operation.placeholder}
                    onFocus={operation.onEnterField}
                    onBlur={operation.onExitField}
                    value={operation.value || value || ''}
                    style={{ marginBottom: '0' }}
                    onChange={(e: any) => operation.setValue(e.target.value)}
                    enterButton={enterButton || t('TableSearchAnt.btnTitle', 'Search')}
                    onSearch={() => searchAction.fire()}
                    onKeyDown={e => this.handleKeyDown(e, operation)}
                    {...props as any}
                />
                {operation.value && operation.value.length > 0 && (
                    <Button
                        id={id + '-clearBtn'}
                        style={clearButtonStyle}
                        htmlType="button"
                        onClick={() => {
                            operation.setValue('');
                            document.getElementById(id)!.focus();
                            searchAction.fire();
                        }}
                    >
                        <Icon type="close" />
                    </Button>
                )}
            </div>
        );
    }
}

@observer
export class TextFormAnt extends React.Component<BindStringAntProps & BindFormItemAntProps> {
    render() {
        const { operation, invisible, ...props } = parsePropsForChild(this.props, this.props.operation);
        if (invisible) {
            return null;
        }
        return (
            <FormItemAnt operation={operation} {...this.props as any}>
                <TextAnt operation={operation} {...props} />
            </FormItemAnt>
        );
    }
}
