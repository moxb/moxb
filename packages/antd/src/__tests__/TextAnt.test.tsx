import { TextImpl, ActionImpl } from '@moxb/moxb';
import { mount, shallow } from 'enzyme';
import * as React from 'react';
import { TextAnt, TextFormAnt, TextSearchAnt } from '../TextAnt';
import { shallowMoxbToJson } from './enzymeHelper';

describe('TextAnt', function () {
    it('should render a text input by default', function () {
        const operation = new TextImpl({
            id: 'TheId',
            label: 'oh',
            disabled: () => false,
        });
        expect(shallowMoxbToJson(shallow(<TextAnt operation={operation} />))).toMatchSnapshot();
    });
    it('should render a textarea with the control option set', function () {
        const operation = new TextImpl({
            id: 'TheId',
            label: 'oh',
            control: 'textarea',
            disabled: () => false,
        });
        expect(shallowMoxbToJson(shallow(<TextAnt operation={operation} />))).toMatchSnapshot();
    });

    it('should return null if invisible', function () {
        const operation = new TextImpl({
            id: 'TheId',
            label: 'oh',
            invisible: () => true,
        });
        expect(shallow(<TextAnt operation={operation} />).type()).toBeNull();
    });

    it('should call onChange if the input value changes', function () {
        const operation = new TextImpl({
            id: 'TheId',
            label: 'oh',
        });
        const wrapper = mount(<TextAnt operation={operation} />);
        const input: any = wrapper.find('input').first();
        input.simulate('change', { target: { value: 'Changed' } });
        expect(wrapper.find('input').first().prop('value')).toBe('Changed');
    });

    it('should call onChange if the textarea value changes', function () {
        const operation = new TextImpl({
            id: 'TheId',
            label: 'oh',
            control: 'textarea',
        });
        const wrapper = mount(<TextAnt operation={operation} />);
        const textarea: any = wrapper.find('textarea').first();
        textarea.simulate('change', { target: { value: 'Changed' } });
        expect(wrapper.find('textarea').first().prop('value')).toBe('Changed');
    });
});

describe('TextFormAnt', function () {
    it('should render form item with a text input by default', function () {
        const operation = new TextImpl({
            id: 'TheId',
            label: 'oh',
            disabled: () => false,
        });
        expect(shallowMoxbToJson(shallow(<TextFormAnt operation={operation} />))).toMatchSnapshot();
    });
    it('should return null if invisible', function () {
        const operation = new TextImpl({
            id: 'TheId',
            label: 'oh',
            invisible: () => true,
        });
        expect(shallow(<TextFormAnt operation={operation} />).type()).toBeNull();
    });
    it('should call onChange if the input value changes', function () {
        const operation = new TextImpl({
            id: 'TheId',
            label: 'oh',
        });
        const wrapper = mount(<TextAnt operation={operation} />);
        const input: any = wrapper.find('input').first();
        input.simulate('change', { target: { value: 'Changed' } });
        expect(wrapper.find('input').first().prop('value')).toBe('Changed');
    });
});

describe('TextSearchAnt', function () {
    it('should render search text input by default', function () {
        const operation = new TextImpl({
            id: 'TheId',
            label: 'oh',
            disabled: () => false,
        });
        const searchAction = new ActionImpl({
            id: 'Action',
            fire: () => {},
        });
        expect(
            shallowMoxbToJson(shallow(<TextSearchAnt operation={operation} searchAction={searchAction} />))
        ).toMatchSnapshot();
    });
    it('should return null if invisible', function () {
        const operation = new TextImpl({
            id: 'TheId',
            label: 'oh',
            invisible: () => true,
        });
        const searchAction = new ActionImpl({
            id: 'Action',
            fire: () => {},
        });
        expect(shallow(<TextSearchAnt operation={operation} searchAction={searchAction} />).type()).toBeNull();
    });
    it('should call onChange if the input value changes', function () {
        const operation = new TextImpl({
            id: 'TheId',
            label: 'oh',
        });
        const searchAction = new ActionImpl({
            id: 'Action',
            fire: () => {},
        });
        const wrapper = mount(<TextSearchAnt operation={operation} searchAction={searchAction} />);
        const input: any = wrapper.find('input').first();
        input.simulate('change', { target: { value: 'Changed' } });
        expect(wrapper.find('input').first().prop('value')).toBe('Changed');
    });
});
