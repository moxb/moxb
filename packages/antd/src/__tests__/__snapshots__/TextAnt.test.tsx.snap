// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextAnt should render a text input by default 1`] = `
<Input
  data-testid="the_id"
  disabled={false}
  id="the_id"
  label="oh"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  readOnly={false}
  type="text"
  value=""
/>
`;

exports[`TextAnt should render a textarea with the control option set 1`] = `
<TextArea
  data-testid="the_id"
  disabled={false}
  id="the_id"
  label="oh"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  readOnly={false}
  value=""
/>
`;

exports[`TextFormAnt should render form item with a text input by default 1`] = `
<FormItemAnt
  operation="bound TextImpl"
>
  <TextAnt
    disabled={false}
    id="the_id"
    operation="bound TextImpl"
    readOnly={false}
  />
</FormItemAnt>
`;

exports[`TextSearchAnt should render search text input by default 1`] = `
<div
  style={
    Object {
      "marginBottom": "1.5em",
      "position": "relative",
    }
  }
>
  <Search
    disabled={false}
    enterButton="Search"
    id="the_id"
    label="oh"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onSearch={[Function]}
    readOnly={false}
    style={
      Object {
        "marginBottom": "0",
      }
    }
    value=""
  />
</div>
`;
